File: top_interview_questions/1_easy/3_linked_list/reverse_linked_list/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# TODO: improve
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        list_vals = []
        while head:
            list_vals.append(head.val)
            head = head.next

        list_vals = list_vals[::-1]
        reversed_list = ListNode(list_vals[0])
        reversed_list_head = reversed_list

        for val in list_vals[1:]:
            reversed_list.next = ListNode(val)
            reversed_list = reversed_list.next

        return reversed_list_head

File: top_interview_questions/1_easy/3_linked_list/reverse_linked_list/solution.go
package main

import "fmt"

func main() {
	ll := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 3,
			},
		},
	}
	ll = reverseList(ll)
	for ll != nil {
		fmt.Print(ll.Val, " ")
		ll = ll.Next
	} // expected 3 2 1
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode
	var next *ListNode

	for head != nil {
		next = head.Next
		head.Next = prev
		prev = head
		head = next
	}

	return prev
}
